WalletWasabi/Wallets/FilterProcessor/ParallelBlockDownloadService.cs

```diff
@@ -0,0 +1,190 @@
+	/// The main use case is to deal with a blockchain reorg.
+	/// </remarks>
+	public void RemoveBlocks(uint maxBlockHeight)
+	{
+		PriorityQueue<Request, uint> tempQueue = new(Comparer<uint>.Default);
+
+		lock (Lock)
+		{
+			int count = BlocksToDownload.Count;
+
+			for (int i = 0; i < count; i++)
+			{
+				if (!BlocksToDownload.TryDequeue(out Request? request, out uint height))
+				{
+					throw new UnreachableException();
+				}
+
+				if (height <= maxBlockHeight)
+				{
+					tempQueue.Enqueue(request, height);
+				}
+			}
+
+			BlocksToDownload.EnqueueRange(tempQueue.UnorderedItems);
+		}
+	}
+
+	/// <summary>
+	/// Downloads blocks in parallel if there are blocks to download.
+	/// </summary>
+	protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+	{
+		try
+		{
+			List<Task<RequestResult>> activeTasks = new(capacity: MaximumParallelTasks);
+
+			while (!cancellationToken.IsCancellationRequested)
+			{
+				bool wait;
+
+				lock (Lock)
+				{
+					// Wait until at least one block-downloading request is here; otherwise, consume requests so that there is at most MAX parallel tasks.
+					wait = activeTasks.Count == 0 && BlocksToDownload.Count == 0;
+				}
+
+				if (wait)
+				{
+					await SynchronizationRequestsSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+				}
+
+				lock (Lock)
+				{
+					int toStart = Math.Min(BlocksToDownload.Count, MaximumParallelTasks - activeTasks.Count);
+
+					for (int i = 0; i < toStart; i++)
+					{
+						// Dequeue does not provide priority value.
+						if (!BlocksToDownload.TryDequeue(out Request? request, out uint blockHeight))
+						{
+							throw new UnreachableException("Failed to dequeue block from the queue.");
+						}
+
+						Task<RequestResult> newTask = Task.Run(
+							async () =>
+							{
+								try
+								{
+									Block? block = await BlockProvider.TryGetBlockAsync(request.BlockHash, cancellationToken).ConfigureAwait(false);
+									return new RequestResult(request.BlockHash, blockHeight, Attempts: request.Attempts + 1, block);
+								}
+								catch (Exception ex)
+								{
+									Logger.LogError($"Exception thrown while getting block {request.BlockHash} (height: {blockHeight})", ex);
+									throw;
+								}
+							},
+							cancellationToken);
+
+						activeTasks.Add(newTask);
+					}
+				}
+
+				if (activeTasks.Count == 0)
+				{
+					continue;
+				}
+
+				Task<RequestResult> task = await Task.WhenAny(activeTasks).ConfigureAwait(false);
+				activeTasks.Remove(task);
+
+				RequestResult result = await task.ConfigureAwait(false);
+
+				// If the block was downloaded OK, we suppose that it's stored on disk and it can be fetched fast.
+				if (result.Block is null)
+				{
+					if (result.Attempts >= MaxFailedAttempts)
+					{
+						Logger.LogInfo($"Attempt to download block {result.BlockHash} (height: {result.BlockHeight}) failed {MaxFailedAttempts} times. Dropping the request.");
```

Comment:
Even if it makes sense to have this retry mechanism, which as stated in my previous comment, I'm not sure. Is failure to get a block really a `LogInfo`? In the usage of this code, what would really happen if we fail to get a block here? Wallet state broken? Or do you defer the responsibility of throwing/logging error to the consumer component?